### PR TITLE
V rchekuri/176819 graph sdk assignments evolvable enums 14 sept

### PR DESCRIPTION
**Pull request by raghuchek**
image

### CODE REVIEW

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 1]
```diff
@@ -64,5 +67,95 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
...
+            try
+            {
+                int retries = 0;
+                string assignmentId = string.Empty;
+                string submissionId = string.Empty;
+
+                // Get a Graph client using delegated permissions
+                var graphClient = MicrosoftGraphSDK.GraphClient.GetDelegateClient(_config["tenantId"], _config["appId"], _config["teacherAccount"], _config["password"]);
+
+                // Create a draft assignment with displayName = "Verify assignments states [inactive]"
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: revise this comment to: "Create assignment to verify inactive state"

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 2]
```diff
@@ -64,5 +67,95 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
...
+                var assignmentInactive = await MicrosoftGraphSDK.Assignment.CreateAsync(graphClient, _config["classId"]);
+                assignmentId = assignmentInactive.Id;
+                Console.WriteLine($"Assignment created successfully {assignmentInactive.Id} in state {assignmentInactive.Status}");
+
+                // Create a draft assignment with displayName = "Verify assignments states [assigned]"
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: revise this comment to: "Create assignment to verify assigned state"

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 3]
```diff
@@ -64,5 +67,95 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
...
+                var assignmentAssigned = await MicrosoftGraphSDK.Assignment.CreateAsync(graphClient, _config["classId"]);
+                Console.WriteLine($"Assignment created successfully {assignmentAssigned.Id} in state {assignmentAssigned.Status}");
+
+                // Create a draft assignment with displayName = "Verify assignments states [draft]"
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: revise this comment to: "Create assignment to verify draft state"

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 4]
```diff
@@ -64,5 +67,95 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
...
+                    // Print . in the log to show that the call is being retried
+                    Console.WriteLine(".");
+
+                    assignmentInactive = await MicrosoftGraphSDK.Assignment.GetAssignmentAsync(graphClient, _config["classId"], assignmentId);
+
+                    // If you are calling this code pattern in Backend agent of your service, then you want to retry the work after some time. The sleep here is just an example to emulate the delay
+                    Thread.Sleep(2000);
+                    retries++;
+                }
+                Console.WriteLine($"Assignment {assignmentInactive.Id} publish is completed. Status: {assignmentInactive.Status}");
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: please move all the PUBLISH code into a new method at the **GlobalMethods** class.
It is a repetitive code across different samples, we can have it in a global mehtod and just call it and pass the assignment object as parameter.

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 5]
```diff
@@ -64,5 +67,95 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
...
+                    // Print . in the log to show that the call is being retried
+                    Console.WriteLine(".");
+
+                    assignmentAssigned = await MicrosoftGraphSDK.Assignment.GetAssignmentAsync(graphClient, _config["classId"], assignmentId);
+
+                    // If you are calling this code pattern in Backend agent of your service, then you want to retry the work after some time. The sleep here is just an example to emulate the delay
+                    Thread.Sleep(2000);
+                    retries++;
+                }
+                Console.WriteLine($"Assignment {assignmentAssigned.Id} publish is completed. Status: {assignmentAssigned.Status}");
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: remove this code and just call the new global method you will create to check the PUBLISH process (from my comment above).

Do the same for any other sample that is publishing assignments.

File: samples/csharp/MicrosoftGraphSDK/Assignment.cs [hunk 6]
```diff
@@ -133,5 +133,64 @@ public class Assignment
...
+                    .GetAsync(requestConfig => {
+                        requestConfig.Headers.Add(
+                            headerName, headerValue);
+                    });
+            }
+            catch (Exception ex)
+            {
+                throw new GraphException($"GetAssignmentsAsync call: {ex.Message}", ex, classId);
+            }
+        }
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftGraphSDK/Assignment.cs]: this method is not used in the sample, remove it please

**Review by cristobal-buenrostro — CHANGES_REQUESTED** [review]

File: samples/csharp/MicrosoftGraphSDK/Assignment.cs (hunk 6 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftGraphSDK/Assignment.cs]: removed the unused method

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 1 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: resolved

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 2 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: resolved

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 3 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: resolved

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 7]
```diff
@@ -64,5 +67,95 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
                 return null;
             }
         }
+        /// <summary>
+        /// Workflow to get assignments evolvable enums
```
Comment (sharad-sharma-msft) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: I almost feel like we need pointers to documentation on evolvable enums , what are those, why they need specific handling etc. we should expand here in terms of comments

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 8]
```diff
@@ -64,5 +67,95 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
...
+                }
+
+                if (assignmentDraft.Status == EducationAssignmentStatus.Draft)
+                {
+                    Console.WriteLine($"Draft Assignment Found: {assignmentId}");
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"AssignmentsEvolvableEnums: {ex.ToString()}");
```
Comment (sharad-sharma-msft) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: nit: separate our words here.

File: samples/csharp/MicrosoftGraphSDK/Assignment.cs [hunk 9]
```diff
@@ -133,5 +133,64 @@ public class Assignment
                 throw new GraphException($"PublishAsync call: {ex.Message}", ex, classId, assignmentId);
             }
         }
+
+        /// <summary>
```
Comment (sharad-sharma-msft) [inline samples/csharp/MicrosoftGraphSDK/Assignment.cs]: add link or pointer to our state machine article for anyone to understand what state transitions are feasible.

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 10]
```diff
@@ -64,5 +67,95 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
                 return null;
             }
         }
+        /// <summary>
+        /// Workflow to get assignments evolvable enums
+        /// </summary>
+        /// <param name></param> 
+        public async Task AssignmentsEvolvableEnums()
```
Comment (sharad-sharma-msft) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: nit: add Async

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 11]
```diff
@@ -64,5 +67,95 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
...
+
+                // Create a draft assignment with displayName = "Verify assignments states [draft]"
+                var assignmentDraft = await MicrosoftGraphSDK.Assignment.CreateAsync(graphClient, _config["classId"]);
+                Console.WriteLine($"Assignment created successfully {assignmentDraft.Id} in state {assignmentDraft.Status}");
+
+                // Teacher publishes the assignment to make it appears in the student's list
+                assignmentInactive = await MicrosoftGraphSDK.Assignment.PublishAsync(graphClient, _config["classId"], assignmentId);
+                Console.WriteLine($"Assignment {assignmentInactive.Id} publish in process");
+
+                // Verify assignment state, publish is completed until state equals "Assigned"
```
Comment (sharad-sharma-msft) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: not completed

File: samples/csharp/MicrosoftGraphSDK/Assignment.cs [hunk 12]
```diff
@@ -133,5 +133,64 @@ public class Assignment
...
+                    });
+            }
+            catch (Exception ex)
+            {
+                throw new GraphException($"DeactivateAsync call: {ex.Message}", ex, classId, assignmentId);
+            }
+        }
+
+        /// <summary>
+        /// Get all the assignments from the class
```
Comment (sharad-sharma-msft) [inline samples/csharp/MicrosoftGraphSDK/Assignment.cs]: nit: for the given class

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 4 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: Updated the method in Global

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 5 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: Updated the method in Global

File: samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs [hunk 13]
```diff
@@ -1,10 +1,15 @@
...
+using MicrosoftGraphSDK;
 
 namespace MicrosoftEduGraphSamples.Utilities
 {
     internal class GlobalMethods
     {
         private static IConfiguration _config;
+        private const int MAX_RETRIES = 10;
 
         public static void ValidateConfiguration(IConfiguration configuration)
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs:18]: I think I missed to ask you for a method header here. Please add it

File: samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs [hunk 14]
```diff
@@ -41,5 +46,27 @@ public static void ValidateConfiguration(IConfiguration configuration)
             }
         }
 
+        public static async Task PublishAssignmentsAsync(Microsoft.Graph.Beta.GraphServiceClient graphClient, string assignmentId)
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs]: add method header

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 15]
```diff
@@ -15,7 +17,8 @@ namespace MicrosoftEduGraphSamples.Workflows
     internal class AssignmentWorkflow
     {
         private readonly IConfiguration _config;
-
+        private const int MAX_RETRIES = 10; // Maximum number of retries for long running operations
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: is this constant still used here?, remove it if not

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 16]
```diff
@@ -64,5 +67,64 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
...
+                // Verifying that you have an Inactive, Assigned and Draft assignments
+                if (assignmentInactive.Status == EducationAssignmentStatus.Inactive)
+                {
+                    Console.WriteLine($"Inactive Assignment Found: {assignmentId}");
+                }
+
+                if (assignmentAssigned.Status == EducationAssignmentStatus.Assigned)
+                {
+                    Console.WriteLine($"Assigned Assignment Found: {assignmentId}");
+                }
```
Comment (cristobal-buenrostro) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: is this log printed when you run this sample locally? can you try and make sure it happens?

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 15 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: removed

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 16 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: yes, I have attached the screen shot previously.

File: samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs (hunk 14 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs]: added

File: samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs (hunk 13 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs:18]: added

File: samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs [hunk 17]
```diff
@@ -41,5 +52,31 @@ public static void ValidateConfiguration(IConfiguration configuration)
...
+
+            // Verify assignment state, publish is completed until state equals "Assigned"
+            while (assignment.Status != EducationAssignmentStatus.Assigned && retries <= MAX_RETRIES)
+            {
+                // Print . in the log to show that the call is being retried
+                Console.WriteLine(".");
+
+                assignment = await MicrosoftGraphSDK.Assignment.GetAssignmentAsync(graphClient, _config["classId"], assignmentId);
+
+                // If you are calling this code pattern in Backend agent of your service, then you want to retry the work after some time. The sleep here is just an example to emulate the delay
```
Comment (sharad-sharma-msft) [inline samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs]: illustrate comment further by naming some of the popular retry policies out there in use like linear, step function, exponential.

**Review by sharad-sharma-msft — CHANGES_REQUESTED** [review]

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 8 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: separated now

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 10 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: added

File: samples/csharp/MicrosoftGraphSDK/Assignment.cs (hunk 12 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftGraphSDK/Assignment.cs]: changed it

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 11 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: added

File: samples/csharp/MicrosoftGraphSDK/Assignment.cs (hunk 9 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftGraphSDK/Assignment.cs]: added a reference

File: samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs (hunk 17 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Utilities/GlobalMethods.cs]: added

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs (hunk 7 above)
Comment (raghuchek) [reply samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs]: added reference

File: samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs [hunk 18]
```diff
@@ -64,5 +66,65 @@ public async Task<IEnumerable<EducationAssignment>> GetMeAssignmentsFromNonArchi
                 return null;
             }
         }
+        /// <summary>
+        /// Workflow to get assignments evolvable enums, Evolvable enums is a mechanism that Microsoft Graph API uses to add new members to existing enumerations without causing a breaking change for applications.
```
Comment (sharad-sharma-msft) [inline samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs:68]: I wonder if you need to wrap the comment

**Review by sharad-sharma-msft — APPROVED** [review]

**Review by sharad-sharma-msft — APPROVED** [review]

**Comment by cristobal-buenrostro** [comment]
you missed to add the sample in the readme file

**Comment by raghuchek** [comment]
> you missed to add the sample in the readme file

pushed the readme changes

File: README.md [hunk 19]
```diff
@@ -116,6 +116,8 @@ Use the `Program.cs` file to test any of the current workflows.
 |1| Reassign submission   | Sample showing how the teacher creates an assignment and then publish it; the student submit his work and then teacher reassign it with feedback.                      |[View](https://github.com/microsoft/edu-assignments-graph-sdk/blob/main/samples/csharp/MicrosoftEduGraphSamples/Workflows/SubmissionWorkflow.cs#L63)|
 |2| Filter archived and deleted classes from assignments | Get me assignments from active and non-archived classes. |[View](https://github.com/microsoft/edu-assignments-graph-sdk/blob/main/samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs#L63)|
```
Comment (cristobal-buenrostro) [inline README.md]: line number for the link must be 
#L28

File: README.md [hunk 20]
```diff
@@ -116,6 +116,8 @@ Use the `Program.cs` file to test any of the current workflows.
...
 |3| Get assignment submissions using batch request   | Since as a teacher you can't expand submissions, you will have to request every assignment for submissions on it. This SDK sample is to illustrate on how you can do this in a single batch request.|[View](https://github.com/microsoft/edu-assignments-graph-sdk/blob/main/samples/csharp/MicrosoftEduGraphSamples/Workflows/SubmissionWorkflow.cs#L154)|
+|4| Assignment evolvable enums   | Evolvable enums is a mechanism that Microsoft Graph API uses to add new members to existing enumerations without causing a breaking change for applications.|[View](https://github.com/microsoft/edu-assignments-graph-sdk/blob/main/samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs#L69)|
```
Comment (cristobal-buenrostro) [inline README.md]: line number for the link must be 
#L70

File: README.md [hunk 21]
```diff
@@ -116,6 +116,8 @@ Use the `Program.cs` file to test any of the current workflows.
 |1| Reassign submission   | Sample showing how the teacher creates an assignment and then publish it; the student submit his work and then teacher reassign it with feedback.                      |[View](https://github.com/microsoft/edu-assignments-graph-sdk/blob/main/samples/csharp/MicrosoftEduGraphSamples/Workflows/SubmissionWorkflow.cs#L63)|
```
Comment (cristobal-buenrostro) [inline README.md]: line number for the link must be 
#L27

File: README.md [hunk 22]
```diff
@@ -116,6 +116,8 @@ Use the `Program.cs` file to test any of the current workflows.
...
 |2| Filter archived and deleted classes from assignments | Get me assignments from active and non-archived classes. |[View](https://github.com/microsoft/edu-assignments-graph-sdk/blob/main/samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs#L63)|
 |3| Get assignment submissions using batch request   | Since as a teacher you can't expand submissions, you will have to request every assignment for submissions on it. This SDK sample is to illustrate on how you can do this in a single batch request.|[View](https://github.com/microsoft/edu-assignments-graph-sdk/blob/main/samples/csharp/MicrosoftEduGraphSamples/Workflows/SubmissionWorkflow.cs#L154)|
```
Comment (cristobal-buenrostro) [inline README.md]: line number for the link must be 
#L103

File: README.md [hunk 23]
```diff
@@ -116,6 +116,8 @@ Use the `Program.cs` file to test any of the current workflows.
...
+|4| Assignment evolvable enums   | Evolvable enums is a mechanism that Microsoft Graph API uses to add new members to existing enumerations without causing a breaking change for applications.|[View](https://github.com/microsoft/edu-assignments-graph-sdk/blob/main/samples/csharp/MicrosoftEduGraphSamples/Workflows/AssignmentWorkflow.cs#L69)|
+|5| Module Workflow   | The Workflow related to modules include creating a draft, publishing, setting up resources, adding link resources, Word document resources, channel resources, assignment resources, patching, and deleting as a teacher.|[View](https://github.com/microsoft/edu-assignments-graph-sdk/blob/main/samples/csharp/MicrosoftEduGraphSamples/Workflows/ModuleWorkflow.cs#L13)|
```
Comment (cristobal-buenrostro) [inline README.md]: line number for the link must be 
#L28

File: README.md (hunk 19 above)
Comment (raghuchek) [reply README.md]: changed

File: README.md (hunk 20 above)
Comment (raghuchek) [reply README.md]: changed

File: README.md (hunk 21 above)
Comment (raghuchek) [reply README.md]: changed

File: README.md (hunk 22 above)
Comment (raghuchek) [reply README.md]: changed

File: README.md (hunk 23 above)
Comment (raghuchek) [reply README.md]: changed

**Review by cristobal-buenrostro — APPROVED** [review]
I made some final edits, but now is good to go. Approved.

**Pull request merged**